runtime/lua/vim/filetype.lua

```diff
@@ -2213,6 +2186,19 @@ local pattern = {
   end,
   -- END PATTERN
 }
+
+-- BEGIN ENV_PATTERN
```

Comment:
Maybe add a comment here how the table works? It's almost-but-not-entirely obvious...

Maybe something like
```
 -- file paths using environment variables: an entry
 -- `["pattern"] = { 'filetype', 'VAR'}
 -- means that files matching the pattern `$VAR/pattern` are detected as `filetype`
 ```